// source-map-js/lib/util.js
/* -*- Mode: js; js-indent-level: 2; -*- */
/*
 * Copyright 2011 Mozilla Foundation and contributors
 * Licensed under the New BSD license. See LICENSE or:
 * http://opensource.org/licenses/BSD-3-Clause
 */

/**
 * This is a helper function for getting values from parameter/options
 * objects.
 *
 * @param args The object we are extracting values from
 * @param name The name of the property we are getting.
 * @param defaultValue An optional value to return if the property is missing
 * from the object. If this is not specified and the property is missing, an
 * error will be thrown.
 */
function getArg(aArgs, aName, aDefaultValue) {
  if (aName in aArgs) {
    return aArgs[aName];
  } else if (arguments.length === 3) {
    return aDefaultValue;
  } else {
    throw new Error('"' + aName + '" is a required argument.');
  }
}
exports.getArg = getArg;

var urlRegexp = /^(?:([\w+\-.]+):)?\/\/(?:(\w+:\w+)@)?([\w.-]*)(?::(\d+))?(.*)$/;
var dataUrlRegexp = /^data:.+\,.+$/;

function urlParse(aUrl) {
  var match = aUrl.match(urlRegexp);
  if (!match) {
    return null;
  }
  return {
    scheme: match[1],
    auth: match[2],
    host: match[3],
    port: match[4],
    path: match[5]
  };
}
exports.urlParse = urlParse;

function urlGenerate(aParsedUrl) {
  var url = '';
  if (aParsedUrl.scheme) {
    url += aParsedUrl.scheme + ':';
  }
  url += '//';
  if (aParsedUrl.auth) {
    url += aParsedUrl.auth + '@';
  }
  if (aParsedUrl.host) {
    url += aParsedUrl.host;
  }
  if (aParsedUrl.port) {
    url += ":" + aParsedUrl.port
  }
  if (aParsedUrl.path) {
    url += aParsedUrl.path;
  }
  return url;
}
exports.urlGenerate = urlGenerate;

var MAX_CACHED_INPUTS = 32;

/**
 * Takes some function `f(input) -> result` and returns a memoized version of
 * `f`.
 *
 * We keep at most `MAX_CACHED_INPUTS` memoized results of `f` alive. The
 * memoization is a dumb-simple, linear least-recently-used cache.
 */
function lruMemoize(f) {
  var cache = [];

  return function(input) {
    for (var i = 0; i < cache.length; i++) {
      if (cache[i].input === input) {
        var temp = cache[0];
        cache[0] = cache[i];
        cache[i] = temp;
        return cache[0].result;
      }
    }

    var result = f(input);

    cache.unshift({
      input,
      result,
    });

    if (cache.length > MAX_CACHED_INPUTS) {
      cache.pop();
    }

    return result;
  };
}

/**
 * Normalizes a path, or the path portion of a URL:
 *
 * - Replaces consecutive slashes with one slash.
 * - Removes unnecessary '.' parts.
 * - Removes unnecessary '<dir>/..' parts.
 *
 * Based on code in the Node.js 'path' core module.
 *
 * @param aPath The path or url to normalize.
 */
var normalize = lruMemoize(function normalize(aPath) {
  var path = aPath;
  var url = urlParse(aPath);
  if (url) {
    if (!url.path) {
      return aPath;
    }
    path = url.path;
  }
  var isAbsolute = exports.isAbsolute(path);
  // Split the path into parts between `/` characters. This is much faster than
  // using `.split(/\/+/g)`.
  var parts = [];
  var start = 0;
  var i = 0;
  while (true) {
    start = i;
    i = path.indexOf("/", start);
    if (i === -1) {
      parts.push(path.slice(start));
      break;
    } else {
      parts.push(path.slice(start, i));
      while (i < path.length && path[i] === "/") {
        i++;
      }
    }
  }

  for (var part, up = 0, i = parts.length - 1; i >= 0; i--) {
    part = parts[i];
    if (part === '.') {
      parts.splice(i, 1);
    } else if (part === '..') {
      up++;
    } else if (up > 0) {
      if (part === '') {
        // The first part is blank if the path is absolute. Trying to go
        // above the root is a no-op. Therefore we can remove all '..' parts
        // directly after the root.
        parts.splice(i + 1, up);
        up = 0;
      } else {
        parts.splice(i, 2);
        up--;
      }
    }
  }
  path = parts.join('/');

  if (path === '') {
    path = isAbsolute ? '/' : '.';
  }

  if (url) {
    url.path = path;
    return urlGenerate(url);
  }
  return path;
});
exports.normalize = normalize;

/**
 * , =1 {Una pestanya} one {# pestanyes} many {# pestanyes} other {# pestanyes}}{NUM_TABS, plural, =1 {$1 - 1 pestanya} one {$1 - # pestanyes} many {$1 - # pestanyes} other {$1 - # pestanyes}}{NUM_TABS, plural, =1 {Grup sense nom - 1 pestanya} one {Grup sense nom - # pestanyes} many {Grup sense nom - # pestanyes} other {Grup sense nom - # pestanyes}}Inicia la sessió per veure les pestanyes d'altres dispositius&HistorialdescarregaS'està obrint a la sessió del navegador existent.Activat: sincronitza-ho totActivat: configuració personalitzadaLa sincronització no funciona.La sincronització de contrasenyes no funcionaPer continuar desant contrasenyes al vostre Compte de Google, verifiqueu la vostra identitatAssegureu-vos que sempre podeu accedir a les dades de sincronitzacióAssegureu-vos que sempre podeu accedir a les contrasenyes desadesObre la configuracióLa teva organització ha inhabilitat la sincronització.S'ha aturat la sincronitzacióConfirma la configuració de sincronització per iniciar la sincronització.Torna-ho a provar. No hem pogut iniciar la sessió.Torna a iniciar la sessió. La sincronització no funciona.La sincronització no funciona. Proveu de tancar la sessió i tornar-la a iniciar.Introdueix la contrasenya per sincronitzar les dadesVerifica la meva identitatTorna a iniciar la sessió per reprendre la sincronitzacióVols traduir la pàgina escrita en $1?Pàgina traduïdaAquesta pàgina no s'ha pogut traduirTradueixTradueix sempre les pàgines que estiguin en $1No tradueixis mai el $1No tradueixis mai aquest llocS'està traduint…Mostra l'originalEls detalls estan amagats mentre compartiu la pantalla{MUTED_NOTIFICATIONS_COUNT, plural,
            =1 {Notificació nova}
            one {# notificacions noves}
            many {# notificacions noves}
            other {# notificacions noves}}{MUTED_NOTIFICATIONS_COUNT, plural,
            =1 {Mostra}
            one {Mostra-ho tot}
            many {Mostra-ho tot}
            other {Mostra-ho tot}}SnoozeS'ha permès l'accés a la ubicacióS'ha denegat l'accés a la ubicacióEl lloc web no pot accedir a la teva ubicació.Bloqueja sempre $1 accedir a la teva ubicacióContinua permetent que aquest lloc accedeixi a la teva ubicacióAquest lloc s'ha bloquejat per accedir a la vostra ubicació.Permet sempre que $1 accedeixi a la vostra ubicacióContinua bloquejant que aquest lloc accedeixi a la teva ubicacióAquest lloc web t'ho tornarà a preguntar la propera vegada.S'ha desactivat la ubicacióS'ha permès el control total dels dispositius MIDIS'ha denegat el control total dels dispositius MIDIAquesta pàgina té un control total dels dispositius MIDI.Bloqueja sempre $1 de tenir el control total dels dispositius MIDI.Continua permetent que aquest lloc tingui el control total dels dispositius MIDIAquest lloc ha estat bloquejat per tenir el control total dels dispositius MIDI.Permet sempre que $1 tingui el control total dels dispositius MIDIContinua bloquejant que aquest lloc tingui el control total dels dispositius MIDIAquesta pàgina té accés a la càmera i al micròfon.Aquesta pàgina no té permís per accedir a la càmera ni al micròfon.Aquesta pàgina té accés al micròfon.Aquesta pàgina està utilitzant la càmera.Aquesta pàgina no té permís per accedir al micròfon.Aquesta pàgina no té permís per accedir a la càmera.Càmera i micròfon permesosCàmera i micròfon bloquejatsEs permet el micròfonCàmera permesaS'ha bloquejat el micròfonCàmera bloquejadaÉs possible que hàgiu de tornar a carregar aquesta pàgina perquè s'apliqui la configuració nova.Canvia aquest paràmetre a la barra d’adreces.Consulta i gestiona les contrasenyes desades al $1Contrasenyes desades per a aquest llocNo s'ha desat cap contrasenya per a aquest llocContrasenyes desades de $1No s'ha desat cap contrasenya per a $1Editeu el nom d'usuariEdita la notaNota: $1Nom d'usuari: $1Inicia la sessió fàcilment als dispositiusInicia la sessió fàcilment$1 inicia la sessió automàticament a alguns llocs amb les contrasenyes que has desat.La teva organització ha desactivat l'accés als fitxers locals de l'ordinadorFitxers d'imatgeFitxers d'àudioFitxers de vídeoFitxers personalitzatsHas activat el mode de pantalla completa.L'extensió "$1" ha disparat la pantalla sencera.Una extensió ha disparat la pantalla sencera.$1 es mostra ara a pantalla sencera.Ara aquesta pàgina es mostra a pantalla completa.$1 es mostra a pantalla sencera i ha inhabilitat el cursor del ratolí.Aquesta pàgina es visualitza a pantalla completa i ha desactivat el cursor del ratolí.$1 ha inhabilitat el cursor del ratolí.Aquesta pàgina ha desactivat el cursor del ratolí.Prem | $1| per mostrar el cursorAquest lloc web utilitza sensors de llum o moviment.Aquest lloc web utilitza sensors de moviment.No es permet que aquest lloc web utilitzi els sensors de llum i moviment.No es permet que aquest lloc web utilitzi els sensors de moviment.Permet sempre que $1 accedeixi als sensorsContinua bloquejant l'accés als sensorsSensors permesosEl sensors s'han bloquejatAquest lloc web està accedint als teus sensors de llum i moviment.Aquest lloc web està accedint als sensors de moviment.Aquest lloc web no té permís per accedir als sensors de llum o moviment.No es permet que aquest lloc web accedeixi als sensors de moviment.Continua permetent l'accés als sensorsBloqueja sempre que $1 accedeixi als sensorsObteniu subtítols per al vostre àudio i vídeoPer obtenir subtítols per al vostre àudio i vídeo, activeu “Subtítols en directe” a la configuracióAixò us permet seleccionar text amb el teclat. Premeu F7 per desactivar-lo.No m'ho tornis a preguntar quan premi F7Vols activar la navegació amb cursor?Permisos de fitxers multimèdia per al "$1""$1" pot llegir i escriure fitxers de vídeo, so i imatges a les ubicacions seleccionades."$1" pot llegir i suprimir fitxers de vídeo, so i imatges a les ubicacions seleccionades.L'extensió "$1" pot llegir fitxers de vídeo, so i imatges a les ubicacions seleccionades.Recomanacions d'ubicacióúltim adjunt el $1Afegeix una ubicacióAfegeix una galeria multimèdia per directoriConnectatNo s'ha connectatSuprimeix definitivament l'accés per a totes les aplicacionsTrieu què voleu compartirTrieu què voleu compartir amb $1Comparteix la pantalla senceraComparteix una finestra de l'aplicacióL'aplicació $1 vol compartir el contingut de la pantalla.El lloc podrà veure el contingut de la pantalla$1 vol compartir el contingut de la vostra pantalla amb $2.Comparteix l’àudio del sistemaComparteix l'àudio de la finestraComparteix l'àudio de la pestanyaCompartiu també l'àudio del sistema. Aquest dispositiu se silenciarà per evitar comentaris.Comparteix també l'àudio del sistemaCompartiu també l'àudio de la pestanyaPer compartir àudio, compartiu una pestanyaPer compartir àudio, compartiu una pestanya o la pantallaPantalla senceraFinestra{SCREEN_INDEX, plural, =1{Pantalla #} one {Pantalla #} many {Pantalla #} other{Pantalla #}}La vostra organització gestiona les opcions d'ús compartit. És possible que alguns elements estiguin amagats.Trieu una altra pantallaSelecciona una altra finestraAquesta pestanyaUna altra pestanyaSeleccioneu una pestanya per a compartirVisualització prèvia de la pestanya compartidaOrigen multimèdia sense nomVoleu permetre $1 veure aquesta pestanya?El lloc podrà veure el contingut d'aquesta pestanyaPermet també l'àudio de la pestanyaHa finalitzat l'ús compartit de la pantallaLa vostra organització no us permet compartir aquest contingut. Si necessiteu ajuda, poseu-vos en contacte amb l'administrador.Aquesta pestanya utilitza la càmera o el micròfonEl contingut d'aquesta pestanya es comparteixAquesta pestanya està reproduint àudio. (Prem Control + M per silenciar-lo)L'àudio d'aquesta pestanya està silenciat. (Prem Control + M per activar el so)Aquesta pestanya està connectada a un dispositiu BluetoothAquesta pestanya està buscant dispositius Bluetooth de forma activa.Aquesta pestanya està connectada a un dispositiu USBAquesta pestanya està connectada a un dispositiu HID.Aquesta pestanya està connectada a un port en sèrieEn aquesta pestanya s'està reproduint un vídeo en mode de pantalla en pantallaAquesta pestanya comparteix la teva pantallaAquesta pestanya està mostrant contingut de realitat virtual en un visor$1: enregistrament amb la càmera o el micròfon$1: contingut de la pestanya compartit$1: el vídeo s'està reproduint en el mode d'imatge en imatge$1: s'està reproduint àudio$1: àudio silenciat$1: dispositiu Bluetooth connectat$1 - Escaneig del Bluetooth actiu$1: dispositiu USB connectat$1: dispositiu HID connectat$1: port en sèrie connectat$1: error de xarxa$1: bloquejat$1: contingut de l'escriptori compartit$1: presentació de contingut d'RV al casc$1 - Forma part d’un grup de pestanyes sense nom$1 - Forma part del grup de pestanyes $2$1 - Ancorat$1 - Permís sol·licitat, premeu F6 per respondreGrup sense nom: $1$1: grup $2ObertS'ha tancatLa pestanya s'ha desplaçat a la dretaLa pestanya s'ha desplaçat a l'esquerraLa pestanya s'ha desplaçat a l'inici de la barra de tabulacionsLa pestanya s'ha desplaçat al final de la barra de tabulacionsLa pestanya s'ha desplaçat a un grup sense nom - $1La pestanya s'ha desplaçat al grup $1 - $2S'ha suprimit la pestanya del grup sense nom - $1S'ha suprimit la pestanya del grup $1 - $2S'està carregant $1$1 - Pestanya inactiva$1 - $2 d’espai lliure$1 - Ús de la memòria - $2$1 - Ús de memòria alt - $2El lloc s'ha actualitzat en segon pla.S'estan cercant els dispositius…Registre del dispositiuAfegeix un paràmetre de consulta a la URL a la pàgina d'actualització automàtica: edge://device-log/?refresh=<sec>Esborra els tipusNo hi ha cap entrada de registreNivell de registre:UsuariIncidènciaDepuraEnergiaBluetoothUSBHIDFIDOSèrieGeolocalitzacióPantallaInformació del fitxerData i hora detallades[$1]
        $2
        $3Go to $1 to connect.S'ha detectat un dispositiu $1Configuració del dispositiu d'interfície humanaQuant als dispositius HIDConfiguració d'USBQuant als dispositius USB{NUM_CONNECTION, plural,
            =0 {L’extensió "{1}" estava accedint a dispositius}
            =1 {L’extensió "{1}" està accedint a {0} dispositiu}
            one {L’extensió "{1}" està accedint a {0} dispositiu}
            many {L’extensió "{1}" està accedint a {0} dispositiu}
            other {L’extensió "{1}" està accedint a {0} dispositiu}}Comunicacions$1 per permetre l'emparellamentActiva el BluetoothObre les preferències del sistemaTorna a explorarTorna a explorar els dispositius Bluetooth$1: emparellatDispositius compatiblesDispositiu desconegut de $1Dispositiu desconegut [$1:$2]Producte desconegut $1 de: $2Producte desconegut $1 del proveïdor $2Dispositius del proveïdor $1Dispositius de $1Dispositius de qualsevol proveïdor$1 vol connectar-se a un port en sèrieConnectaS'estan cercant dispositius sèrie…Dispositiu USB de $2 (product $1)Dispositiu USB ($2:$1)Dispositius USB del proveïdor $1Dispositius USB de $1Qualsevol port en sèrie$1 vol connectar-se a un dispositiu HIDDispositiu desconegut ($1)S'estan cercant dispositius HID…Dispositiu HID ($1:$2)Dispositius HID del proveïdor $1Di
